### PR TITLE
add PolyhedronConstraint as a kinematic constraint.

### PR DESCRIPTION
**Pull request by hongkai-dai**
This is motivated by the stackoverflow question https://stackoverflow.com/questions/72999101/slightly-different-angle-between-vectors-constraint and https://stackoverflow.com/questions/72984441/general-cone-gaze-constraint, both can be modelled with this PolyhedronConstraint.

cc @RussTedrake since you answered the stack overflow question.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@aykut-tri for feature review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596#-:-N7Is52N5KuUXSE8xw1O:b-w629gj)* status: LGTM missing from assignee aykut-tri, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @aykut-tri)

**Review by aykut-tri — COMMENTED** [review]
The feature :lgtm_strong: with only a style nit.

+@ggould-tri for platform review please, thanks!

> Reviewed 9 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596#-:-N7NORuF3OiA3wH1GvZS:b-2gan9a)* status: LGTM missing from assignee ggould-tri(platform) (waiting on @ggould-tri)

___
*[`multibody/inverse_kinematics/polyhedron_constraint.h` line 33 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17596#-N7M9XDSBCSSUKxAKnpu:-N7M9XDT7azgFfXMX1Be:bpg2mjg) ([raw file](https://github.com/robotlocomotion/drake/blob/67e547c60744f62d37e053339b91813f4802271c/multibody/inverse_kinematics/polyhedron_constraint.h#L33)):*
> ```objc
>    * @param A We impose the constraint A * [p_FP1; p_FP2; ...; p_FPn] <= b. @pre
>    * A.cols() = 3 * p_GP.cols();
>    * @param b We impose the constraint A * [p_FP1; p_FP2; ...; p_FPn] <= b
> ```

FYI, the indentations are different, e.g., between `A.cols() = 3 * p_GP.cols();` and `'plant'.  We will update ...`



_Code quote:_
```objc
   * @param A We impose the constraint A * [p_FP1; p_FP2; ...; p_FPn] <= b. @pre
   * A.cols() = 3 * p_GP.cols();
   * @param b We impose the constraint A * [p_FP1; p_FP2; ...; p_FPn] <= b
   * @param plant_context The Context that has been allocated for this
   *   `plant`.  We will update the context when evaluating the constraint.
   *   `plant_context` should be alive during the lifetime of this constraint.
```

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596)* status: LGTM missing from assignee ggould-tri(platform) (waiting on @ggould-tri)

___
*[`multibody/inverse_kinematics/polyhedron_constraint.h` line 33 at r1](https://reviewable.io/reviews/robotlocomotion/drake/17596#-N7M9XDSBCSSUKxAKnpu:-N7NeguzEUztL7vKhWTP:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/67e547c60744f62d37e053339b91813f4802271c/multibody/inverse_kinematics/polyhedron_constraint.h#L33)):*
<details><summary><i>Previously, aykut-tri (Aykut Onol) wrote…</i></summary><blockquote>

FYI, the indentations are different, e.g., between `A.cols() = 3 * p_GP.cols();` and `'plant'.  We will update ...`
</blockquote></details>

Done.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 2 of 9 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596)* status: 2 unresolved discussions, LGTM missing from assignee ggould-tri(platform) (waiting on @aykut-tri and @hongkai-dai)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/17596#-N7QtBwU3_UyNewsvAYe:-N7QtBwU3_UyNewsvAYf:b-9q7cy6) (no related file):*
Since this is motivated by the StackOverflow questions, would it be possible to include an example (probably in the test file?) of how it can be used to solve each of their problems, i.e. to write a gaze constraint or a projected angle constraint?  That way the answers to their questions would live in git rather than just in StackOverflow.

___
___
*[`multibody/inverse_kinematics/polyhedron_constraint.h` line 15 at r2](https://reviewable.io/reviews/robotlocomotion/drake/17596#-N7Qrq75-XDhjxcn2hPD:-N7Qrq75-XDhjxcn2hPE:bw12q38) ([raw file](https://github.com/robotlocomotion/drake/blob/4d93a073b5b738f351fffedf7cd568f83adb3bd4/multibody/inverse_kinematics/polyhedron_constraint.h#L15)):*
> ```objc
>  * measured and expressed in frame F. Notice the constraint is imposed on the
>  * stacked column vector [p_FP1; p_FP2; ...; p_FPn], not on each individual
>  * constraint.
> ```

minor:  You seem to be using the word "constraint" twice in this sentence for different meanings ("the constraint is [not] imposed on... each individual constraint").  Can you reword this to clarify what "each individual constraint" you mean?

(Possibly you meant "point" instead of "constraint"?)



_Code quote:_
```objc
constraint.
```

**Review by ggould-tri — COMMENTED** [review]
:lgtm:

> Reviewed 4 of 9 files at r1, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596#-:-N7VnWfsDKvcLAGuYN9X:bnfp4nl)* status: :shipit: complete! all discussions resolved, LGTM from assignees aykut-tri,ggould-tri(platform) (waiting on @hongkai-dai)

___
*[`multibody/inverse_kinematics/test/polyhedron_constraint_test.cc` line 185 at r3](https://reviewable.io/reviews/robotlocomotion/drake/17596#-N7VnHR_E-Eg8fFRUKeJ:-N7VnHR_E-Eg8fFRUKeK:bn335fi) ([raw file](https://github.com/robotlocomotion/drake/blob/107dc80a7c05ad699302204f3cb64651c43f9125/multibody/inverse_kinematics/test/polyhedron_constraint_test.cc#L185)):*
> ```cpp
>   dut.Eval(q, &y);
>   EXPECT_FALSE((y.array() <= dut.upper_bound().array() + 1E-10).all());
> }
> ```

BTW:  Excellent; thank you!  I think these will be of great help to users.

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 1 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/17596)* status: :shipit: complete! all discussions resolved, LGTM from assignees aykut-tri,ggould-tri(platform) (waiting on @hongkai-dai)
